File: compare.js
/*
JS

- Loosely-typed language
- Functional
- Interpreted at runtime (V8 / node.js)
- Object
- Prototype
- Class (ES6)
*/

var i = 0;
let j = 0;

j = "foo"; // string

// Javascript with parentheses and semicolons
for (let i = 0; i < 10; i++) {
    console.log(i);
}

// Javascript assignment
var foo = "bar";


let foo = "bar";


bar = "blah";



// Javascript export
const Bar = () => {};

module.exports = {
  Bar
}


// Javascript import
var foo = require('foo');
foo.bar();



// Javascript - return multiple values
function foo() {
    return {a: 1, b: 2};
}
const { a, b } = foo();


// Node error handling
foo('bar', function(err, data) {
    //handle error

    if (err) console.log(err);
});


// Variadic function
function foo (...args) {
    console.log(args.length);
}

foo(); // 0
foo(1, 2, 3); // 3

File: compare.go
/*
Go / Golang

- Strong-typed
- Compiled to machine code
- Not OO
- Struct
- Interface
- Composition
*/

var i int

i = "fooo" // compile err

var s string




// Go uses minimal punctuation
for i := 0; i < 10; i++ {
    fmt.Println(i)
}

// Go assignment
var foo string //without type derivation
foo = "bar"

foo := "bar" //with type derivation


bar = "blah"


// Go export
package apackagename //define package name


func Bar (s string) string {
    //Bar will be exported
}

func bar (s string) string {

}

bool 
int


// Go import
import (
    "fmt" // part of Go’s standard library
    "github.com/foo/foo" // imported directly from repository
)
foo.Bar()


// Go - return multiple values
func bar() (int, error) {
    return 1, nil
}
a, b := bar()


//Go error handling
foo, err := bar()
if err != nil {
    // handle error with defer, panic, recover, or log.fatal, etc...
}


// Variadic function
func foo (args ...int) {
	fmt.Println(len(args))
}

func main() {
	foo() // 0
	foo(1,2,3) // 3
}

